### PR TITLE
Faster point cloud rendering by caching decoded data blocks

### PR DESCRIPTION
**Pull request by uclaros**
## Description
When rendering point cloud data, QGIS spends a lot of time decompressing the same laz node data over and over again while panning or zooming.
With this PR, the decompressed points are cached for each combination of _layer uri_, _node_, _requested attributes_, _filter extent_ and _filter expression_, allowing for very fast rendering of cached nodes.
A fixed size cache of 200MB is used and shared across all point cloud layers and canvases (2d, 3d, profile)

Performance can vary depending on actual data, but on a local sample copc file I got render times dropping from ~2.7sec to ~0.35sec once the requested nodes got cached.

### CODE REVIEW

File: src/core/pointcloud/qgspointcloudclassifiedrenderer.cpp [hunk 1]
```diff
@@ -149,16 +149,13 @@ void QgsPointCloudClassifiedRenderer::renderBlock( const QgsPointCloudBlock *blo
...
-      else
-      {
-        const double size = pointSizes.value( attributeValue );
-        drawPoint( x, y, color, size, context );
-        if ( renderElevation )
-          drawPointToElevationMap( x, y, z, size, context );
-      }
+
+      const double size = pointSizes.value( attributeValue );
+      drawPoint( x, y, color, size, context );
```
Comment (nyalldawson) [inline src/core/pointcloud/qgspointcloudclassifiedrenderer.cpp:155]: How's this work with semi transparent triangles? There's a mechanism in the tiled scene renderer for this kind of temporary results which should be used instead

Comment (uclaros) [reply src/core/pointcloud/qgspointcloudclassifiedrenderer.cpp:155]: You're right, large points are also visible around the triangles' outer edges as in the failing tests.
I'm reverting this change and properly handle that on a separate PR as you suggest.

Comment (uclaros) [reply src/core/pointcloud/qgspointcloudclassifiedrenderer.cpp]: Added this on #56134

File: src/core/pointcloud/qgspointcloudrequest.cpp [hunk 2]
```diff
@@ -21,6 +21,12 @@
 
 QgsPointCloudRequest::QgsPointCloudRequest() = default;
 
+bool QgsPointCloudRequest::operator==( const QgsPointCloudRequest &other ) const
+{
+  return mFilterRect == other.filterRect() &&
+         mAttributes.toFields() == other.attributes().toFields(); //todo: QgsPointCloudAttributeCollection::operator==
```
Comment (wonder-sk) [inline src/core/pointcloud/qgspointcloudrequest.cpp:27]: remove todo?

File: src/core/pointcloud/qgsremotecopcpointcloudindex.cpp [hunk 3]
```diff
@@ -111,11 +118,18 @@ std::unique_ptr<QgsPointCloudBlock> QgsRemoteCopcPointCloudIndex::nodeData( cons
     QgsDebugError( QStringLiteral( "Error downloading node %1 data, error : %2 " ).arg( n.toString(), blockRequest->errorStr() ) );
   }
 
+  storeNodeDataToCache( block.get(), n, request );
```
Comment (wonder-sk) [inline src/core/pointcloud/qgsremotecopcpointcloudindex.cpp:121]: looks like we're storing to cache even if the block failed to download - is that ok?

File: src/core/pointcloud/qgscopcpointcloudindex.cpp [hunk 4]
```diff
@@ -154,17 +159,19 @@ std::unique_ptr<QgsPointCloudBlock> QgsCopcPointCloudIndex::nodeData( const Inde
...
   {
-    QgsDebugError( QStringLiteral( "Could not read file %1" ).arg( mFileName ) );
+    QgsDebugError( QStringLiteral( "Could not read file %1" ).arg( mUri ) );
     return nullptr;
   }
   QgsRectangle filterRect = request.filterRect();
 
-  return QgsLazDecoder::decompressCopc( rawBlockData, *mLazInfo.get(), pointCount, requestAttributes, filterExpression, filterRect );
+  std::unique_ptr<QgsPointCloudBlock> decoded = QgsLazDecoder::decompressCopc( rawBlockData, *mLazInfo.get(), pointCount, requestAttributes, filterExpression, filterRect );
+  storeNodeDataToCache( decoded.get(), n, request );
```
Comment (wonder-sk) [inline src/core/pointcloud/qgscopcpointcloudindex.cpp:173]: ...and here (and also in ept point cloud index)

File: src/core/pointcloud/qgspointcloudblock.h [hunk 5]
```diff
@@ -45,6 +45,12 @@ class CORE_EXPORT QgsPointCloudBlock
     //! Dtor
     virtual ~QgsPointCloudBlock() = default;
 
+    /**
+     * Clones the QgsPointCloudBlock returning a new copy.
+     * Caller takes ownership of the returned object.
+     */
+    QgsPointCloudBlock *clone() const;
```
[the remark this answers is not in the document]
Comment (uclaros) [reply src/core/pointcloud/qgspointcloudblock.h]: true

File: src/core/pointcloud/qgspointcloudrequest.cpp (hunk 2 above)
Comment (uclaros) [reply src/core/pointcloud/qgspointcloudrequest.cpp]: yep

File: src/core/pointcloud/qgsremotecopcpointcloudindex.cpp (hunk 3 above)
Comment (uclaros) [reply src/core/pointcloud/qgsremotecopcpointcloudindex.cpp:121]: It doesn't make sense to cache them, especially the remote ones.
I've added a check within `storeNodeDataToCacheStatic` to handle all cases.

**Review by wonder-sk — APPROVED** [review]

**Pull request merged**